### ISSUE TITLE
[ConditionalFact] or [ActiveIssue]

### ISSUE THREAD

**Issue by zhenlan** [dotnet/wcf#1034]
Both ConditionalFact and ActiveIssue can be used to skip tests. There are some pros, cons and expectations of each approach. Start this issue to discuss what we should use going forward.

**Comment by roncain** [comment]
Good idea.  To open the discussion, it is important to understand some concepts:
- [ActiveIssue] is based on xunit trait framework
- [ConditionalFact] is based on the xunit [Fact] base class

Traits are logically part of the "discovery" phase, and [Fact] is logically part of the "execute" phase.

When you use "-notrait" to exclude a group of tests, the xunit discovery phase just throws them away.  You don't see any record they even existed or were skipped.

When you use [ConditionalFact], xunit's [Fact] framework gives it a chance to say "now is your chance to run this test and tell me what result you want, including an indication of whether it failed or was skipped".  If we want a test run to log which tests were skipped and why, some kind of [Fact]-based approach is probably the best choice.

One option we should consider is writing our own form of [ConditionalFact] that allows us to associate a test with a specific issue (or set of issues).  We could also allow control over whether we wanted the test skipped, executed as is, or failed for a different reason, such as "test X failed because attempting to open the firewall for port XYZ failed".  Ideas welcome -- those are just some of the background thoughts.

**Comment by zhenlan** [comment]
This is what was discussed:
- It is preferred that the number of tests that are skipped can be shown as part of test result, which is what [ConditonalFact] can give.
- However, unlike [ActiveIssue], [ConditionalFact] cannot be applied to a test multiple times to indicate different issues on different platforms etc.
- We basically need a new attribute that has the pros of both, but we believe it's hard to change xunit engine.
- **Plan**: we will prototype a custom solution something like below and add it to buildtools repo if it works.
```
  [CondtionalFact("NoIssuesApplicable")]
  [Issue(123, PlatformID.Windows)]
  [Issue(456, PlatformID.OSX)]
  public static void SomeOfMyTest()
  {
	...
  } 
```

**Comment by roncain** [comment]
After prototyping this and experimenting, I've altered the approach slightly.  

The design approach is:

- Replace use of ConditionalFact with a new Fact attribute that can reflect over other WCF-specific attributes to determine whether the test should be run or skipped
- Create new "skip" attribute types to allow skipping based on (1) an active issue, (2) some condition (like ConditionalFact), or (3) the current platform. 

Reasons for the approach:

- Choosing a new Fact attribute allows us greater control, not only for skipping tests but doing things like reporting which tests were run, optionally allowing TestProperties to alter behavior, etc.
- We now have greater flexibility in adding new kinds of special-behavior attributes.
- Tests with active issues or platform-specific requirements can just say that with attributes.  They don't need to include some other magic condition member to make them work.
- By implementing our own Condition attribute, we have greater control over how we extend it

The implementation is:

- Create new WcfFact and WcfTheory attributes and discoverers
- Create new IssueAttribute, ConditionAttribute, and PlatformAttribute for the "skip" attributes
- Make the new WcfFact and WcfTheory discovers reflect over their test method for these new attributes and invoke them to determine whether or not a test should be skipped.
- Naturally, all the new skip attributes can be combined or omitted as appropriate.  When there are multiple, evaluation precedence is (1) Platform, (2) Issue, (3) Condition.  This permits tests to be skipped due to Platform or Issues prior to running any Condition code (that could have side-effects such as loading certs)

The proposal to move forward is:

- Check in the proposed changes in one PR (it is non-invasive of existing code)
- Sweep codebase and change all Fact WcfFact and Theory to WcfTheory
- Convert all existing ConditionalFactAttribute to just be WcfFactAttribute + ConditionAttribute

Sample usage is:

```
    [WcfFact]
    [Condition(nameof(Is_WeekEnd))]
    [OuterLoop]
    public static void Skipped_Due_To_False_Condition()
    {
    }

    [WcfFact]
    [Issue(999)]
    public static void Skipped_Due_To_Issue_All_Platforms
    {
    }

    [WcfFact]
    [Issue(999, PlatformID.AnyUnix)]
    public static void Skipped_Due_To_Issue_On_All_Unix_Platforms
    {
    }

    [WcfFact]
    [Platform(PlatformID.AnyUnix)]
    public static void Runs_Only_On_Any_Unix_OS()
    {
    }

    [WcfFact]
    [Platform(PlatformID.AnyUnix, Exclude=PlatformID.OSX)]
    public static void Runs_On_Any_Unix_OS_Except_OSX()
    {
    }
```

**Comment by zhenlan** [comment]
It's probably because of underlying implementation detail, but it feels confusing to me when FrameworkAttribute, OsAttribute or PlatformAttribute should be used. I am wondering what's the scenario we want to disable a particular platform/OS without saying anything. Wouldn't IssueAttribute suffice for what we need?

**Comment by roncain** [comment]
The main purpose of PlatformAttribute, FrameworkAttribute and OSAttribute are to condition tests that should run only on specific Platforms, Frameworks or OSes -- or to *not* run.  They would be used instead of an IssueAttribute in scenarios where it is by-design, not an issue. For example, the WebSockets tests all fail under Windows 7, and I would like to say something like:

```
  [OS(Exclude=OSID.Windows7)]
```

Similarly, if some test could only run on NetCore and not against the full framework, I might say:

```
  [Framework(FrameworkID.NetCore)]
```

If we used IssueAttribute for this, we would need to open issues (and close them?) to represent them, even if we had no intention of fixing that issue.

If you prefer, we could replace these with a single attribute like this:

```
  [Supported(Framework=FrameworkID.NetCore, OS=OSID.AnyUnix)]
```

**Comment by roncain** [comment]
Incidentally, the separate attributes had an "Exclude" optional parameter which was useful to say "don't run on X".  If we move to a SupportAttribute, we would probably want to have parameters for ExcludePlatform, ExcludeFramework and ExcludeOS to make that easy, like:

```
 [Supported(ExcludeOS = OSID.Windows7)]
```

**Comment by roncain** [comment]
Fixed by PR #1429
